mlir/include/mlir/Dialect/LLVMIR/LLVMAttrs.h

```diff
@@ -23,15 +23,24 @@
 namespace mlir {
 namespace LLVM {
 
-/// This class represents the base attribute for all debug info attributes.
-class DINodeAttr : public Attribute {
+/// This class represents the base attribute for all metadata attributes.
+class MDNodeAttr : public Attribute {
```

Comment:
Can you explain why this new base class is necessary? It seems like the newly introduced attributes are also debug info attributes and I would expect they are derived from DINodeAttr.

Reply:
DIExpression and DIGlobalVariableExpression inherit MDNode directly instead of DINode in LLVM. So I felt like needed to introduce a new base attribute class representing MDNode in order to make things work in the translator and maintain consistency with the LLVM side of things.

Reply:
Ah I see, MDNode sounds like it is any metadata node? It feels like Attribute is the direct counterpart for it in MLIR (and we probably don't want to replicate it)? DILocation, which also derives from MDNode, is translated with a separate function in the debug info translation so maybe this would be the right approach for DIGlobalVariableExpression and DIExpression?

Reply:
`Ah I see, MDNode sounds like it is any metadata node?`

MDNode is the base for all debug information in LLVM (https://llvm.org/doxygen/classllvm_1_1Metadata.html).

` It feels like Attribute is the direct counterpart for it in MLIR (and we probably don't want to replicate it)?`

Maybe. I implemented it this just to follow convention without breaking what was there before.

`DILocation, which also derives from MDNode, is translated with a separate function in the debug info translation so maybe this would be the right approach for DIGlobalVariableExpression and DIExpression?`

I feel like this is a slippery slope (you let one slide, then you have to let them all slide). The implementation of DILocation seemed like a means to an end as it probably would've been the only one-off at the time of implementing DINode.

```
MDNodeAttr translate(llvm::DINode *node) {
  return translate(cast<llvm::MDNode>(node));
}
```

I can template this method and, in theory, any debug information could be supported.

Reply:
I have seen the class diagram and i definitely don't like the name `MDNode` isn't that the shortcut for metadata node? If we want to replicate, I would at least call it DIBaseAttr or something similar. There is a lot of other metadata and `MDNode` sound very confusing to me. 

The separate translation function for the location makes sense since there is a Location class in MLIR that does not map to DILocation directly and it is a top-level class that is then attached to an operation/instruction. 

To me it seems like DIGlobalVariableExpression is only used by operations / intrinsics? That means having a function that returns a DIGlobalVariableExpression and not an MDNode base class seems beneficial for such a class. Or is there a use case where it is used in a polymorphic manner, i.e.,  another debug node that has a field which may be any kind of MDNode? If the later exists then it indeed makes sense to have a translate method that returns an MDNodeAttr. Otherwise, I am not yet convinced :).

Reply:
If the priority isn't having a single public API call to translate all things MDNode/DINode then I'll revert back to the way it was as this doesn't impact my implementation whatsoever. It was convenient that translate() already handled most of the things and I just did the work to take it the extra mile to handle more of them.

Reply:
Maybe give other reviewers a chance to comment before changing. I would say if the base class (MDNode) is not really used as a polymorphic field in another debug node then having a separate function makes sense.

Reply:
I do understand both sides of this argument. Reflecting LLVM's class hierarchy seems natural and helpful. On the other hand, there are tons of metadata nodes that we model and non of them would inherit from this.
I suggest to maybe consider adding this in a followup, instead of this revision.

Reply:
I'll revert this. DIGlobalVariable is compatible with the way it was so I'll implement it like DILocalVariable.